### DISCUSSION TITLE
Feedback to the plan for Torch integration to ML.NET

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [xamarin/TorchSharp#335] [General]
[TorchSharp integration to ML.NET](https://github.com/dotnet/machinelearning/projects/26#card-62428046)is currently being planned (supervised by @ericstj ) as recently shared/discussed by @NiklasGustafsson 

I suggest TorchSharp users provide feedback to the integration process here.

As part of the integration, the question h[ow to consume exported Torch model by ML.NET](https://github.com/dotnet/machinelearning/projects/26#card-62428098) has been planned.

I also would like to raise some issues raised before related to the above.
- Will there be an improvement in how TorchSharp consume Torch model created in python. Currently we need an additional python script to create a TorchSharp compatible model in python. Is there a plan soon that **this additional step will be removed**, perhaps as part of the ML.NET integration planning process? 
- There are challenges dealing with ONNX, especially when dealing with NLP use cases (e.g. [ref 1](https://github.com/dotnet/machinelearning/issues/5874#issuecomment-889848670), [ref 2](https://github.com/dotnet/machinelearning/issues/5874#issuecomment-900495155)). Is there a plan to consider incorporating [**ONNX save and load features in TorchSharp**](https://github.com/xamarin/TorchSharp/issues/146#issue-629388982)? By having Onnx save/load in TorchSharp, this provides an alternative path to the existing option to import ONNX to ML.NET

**Comment by ericstj** [comment]
cc @michaelgsharp @briacht
At the moment we're still trying to understand what this work would look like.  Thanks for raising concerns and we'll make sure to stay engaged in the discussion.

**Comment by GeorgeS2019** [comment]
FYI: **latest** [plan for Deep learning and TorchSharp integration to ML.NET!!!](https://github.com/dotnet/machinelearning/issues/5918#issue-983175573)

**Comment by GeorgeS2019** [comment]
Currently it is possible in ML.NET to import Tensorflow trained model file .pb
![image](https://user-images.githubusercontent.com/49812372/136858848-332ec873-679b-4dd8-be28-c28dea723776.png)

ML.NET needs native binding to load PyTorch trained model file e.g. pth.  TorchSharp is key to sort out the paths how ML.NET will load such trained file.

The LibTorch is able to load different formats of PyTorch trained model files
- PyTorch `torch.save` model file: savefile.**pth**
- TorchScript [` torch.jit.save`](https://pytorch.org/docs/stable/generated/torch.jit.save.html) model file:  savefile.**pt**

## Question
Shall ML.NET support one of the two or both file formats?

**Comment by GeorgeS2019** [comment]
***.pth** and ***.data** [Torchsharp specific format for import and export] discussed in [saveload.md](https://github.com/dotnet/TorchSharp/blob/main/docfx/articles/saveload.md)

With this new [implementation](https://github.com/dotnet/TorchSharp/pull/644), now ***.pt** is introduced to the saveload.md

[Discussion ](https://github.com/dotnet/TorchSharp/pull/644#issuecomment-1172977753)on how to integrate TorchSharp to ML.NET
